### ISSUE TITLE
[DISCUSSION] CI Test & Deploy

### ISSUE THREAD

**Issue by mli** [apache/incubator-mxnet#5321]
We recently moved some tests into a jenkins cluster with GPU test enabled. This issue discusses future improvements for both CI test and deploy. 

# What to test

The tests can be categorized into four aspects:

1. **Sanity check** such as lint which checks the code style 
2. **Unit test**: test individual functions and classes, such as testing operators.
3. **Integration test**: test the correctness of multi-modules. For example, training a CNN model on mnist, or predicting on a pre-trained model. We can reuse the codes on folder `example/`
4. **Performance test**: Since we care about system performance, we should benchmark the performance from time to time.

We have test codes for each category somewhat, but still away from 100% test coverage. I roughly list some issues here:

1. lint: We have roughly 40K lines python codes, but only `python/mxnet` is checked. And we have put too many global `pylint: disable=` on the beginning of each file. #5307 is trying to improve it.
2. unittest: We should enable `docstring test` for python, so a same piece of codes can serve as both document and unittest. Also consider to enable pair-test with other libraries such as caffe or torch. 
3. integration/performance test: There are several `test_*.py` scripts on `example/` and `tools/`. But they are not been automatically tested.
 
# How to test 

Each category has different time cost:

1. ~1min
2. 10min ~ 30min
3. \>1h
4. \>1h

Therefore we can run category 1 & 2 for every pull request, but let 3 & 4 run periodically, such every night.  

Besides, we need to test on various environments, including hardware such as CPU/GPU, OSs such as linux, mac os x, windows, android, and libraries such as CUDA 7.5, CUDA 8, MKL-DNN, ...
It is unlikely we can cover all combinations, so we should focus on common cases:

a. Linux (Ubuntu 14.04, 16.04, Centos) with CPU (openblas, MKLDNN) and GPU (CUDA 7.5, CUDA 8)
b. Mac os x with CPU
c. Windows with CPU

We can cover a. with jenkins and docker on AWS. b. is running on travis while c. is running on appveyor. We should move both travis and appveyor tasks into jenkins as well if we figured out how to do it. (Ly is working on the latter)

A sample unittest build contains several subtasks:

1. start a docker instance with the desired environment set up (<1min)
2. compile (3min - 10min)
3. c++ unittest (<1min)
4. python unittest (<5min)
5. scala unittest (<5min)
6. r unittest (<5min)
7. ...

We can run these subtasks in parallel if a single build takes more than 30min.

# Deploy

We should provide pre-build library so that a user can use it directly instead of building from scratch. Currently we host pip build on travis, which supports mac os x cpu, linux cpu & gpu. We should add windows and android supports in the future.

**Comment by mli** [comment]
@lxn2 @kevinthesun @piiswrong

**Comment by kevinthesun** [comment]
Also should we test multi-GPU case? Currently our testing ec2-instance only supports single gpu.

**Comment by eric-haibin-lin** [comment]
I don't see `LCOV - code coverage report` enabled on our Jenkins. It would be nice to have it. Then we can pinpoint the folders not thoroughly tested.

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
